caravel/tasks.py

```diff
@@ -0,0 +1,160 @@
+        updated_sql += sql_statement + "\n"
+    return updated_sql
+
+
+def get_session():
+    """Creates new SQLAlchemy scoped_session."""
+    engine = create_engine(
+        app.config.get('SQLALCHEMY_DATABASE_URI'), convert_unicode=True)
+    return scoped_session(sessionmaker(
+        autocommit=False, autoflush=False, bind=engine))
+
+
+@celery_app.task
+def get_sql_results(database_id, sql, user_id):
+    """Executes the sql query returns the results.
+
+    :param database_id: integer
+    :param sql: string, query that will be executed
+    :param user_id: integer
+    :return: dataframe, query result
+    """
+    # Create a separate session, reusing the db.session leads to the
+    # concurrency issues.
+    query_session = get_session()
+    try:
+        db_to_query = query_session.query(models.Database).filter_by(
+            id=database_id).first()
+    except Exception as e:
+        return {'error': utils.error_msg_from_exception(e)}
+
+    if not db_to_query:
+        return {
+            'error': "Database with id {0} is missing.".format(database_id)}
+
+    query = init_query(database_id, sql, user_id, int(time.time() * 1000))
+    query_session.add(query)
+    query_session.commit()
+
+    query_result, success = get_sql_results_internal(sql, db_to_query)
+
+    query.end_time = int(time.time() * 1000)
+    if success:
+        query.query_status = models.QueryStatus.FINISHED
+        # TODO(bkyryliuk): fill in query.tmp_table_name
+    else:
+        query.query_status = models.QueryStatus.FAILED
+
+    query_session.commit()
+    # TODO(bkyryliuk): return the tmp table  / query_id
+    return query_result
+
+
+# TODO(bkyryliuk): merge the changes made in the carapal first
+#                   before merging this PR.
+def get_sql_results_internal(sql, db_to_query):
+    """Get the SQL query resulst from the give session and db connection.
+
+    :param sql: string, query that will be executed
+    :param db_to_query: models.Database to query
+    :return: (dataframe, boolean), results and the status
+    """
+    try:
+        eng = db_to_query.get_sqla_engine()
+        sql = sql.strip().strip(';')
+        # TODO(bkyryliuk): fix this case for multiple statements
+        if app.config.get('SQL_MAX_ROW'):
+            qry = (
+                select('*')
+                .select_from(TextAsFrom(text(sql), ['*']).alias('inner_qry'))
+                .limit(app.config.get('SQL_MAX_ROW'))
+            )
+            sql = str(qry.compile(eng, compile_kwargs={"literal_binds": True}))
+        if db_to_query.select_as_create_table_as:
+            # TODO(bkyryliuk): figure out if the query is select query.
+            sql = create_table_as(
+                sql, "query_" + str(int(round(time.time() * 1000))))
+
+        df = pd.read_sql_query(sql=sql, con=eng)
+        data = {
+            'columns': [c for c in df.columns],
+            'data': df.to_dict(orient='records'),
+        }
+        return data, True
```

Comment:
I'm not huge on functions that return tuples, but I get what you're doing here. I'd rather add a `success` key to that dict I think.